### PR TITLE
Simplify dynamic filter matching in plans

### PR DESCRIPTION
**Pull request by raunaqmorarka**

### CODE REVIEW

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java [hunk 1]
```diff
@@ -254,8 +261,8 @@ public void testNestedDynamicFiltersRemoval()
                                 INNER,
                                 ImmutableList.of(equiJoinClause("ORDERS_CK", "ORDERS_CK6")),
                                 ImmutableMap.of("ORDERS_CK", "ORDERS_CK6"),
-                                Optional.empty(),
-                                tableScan("orders", ImmutableMap.of("ORDERS_CK", "clerk")),
+                                anyTree(
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:264]: `testNestedDynamicFiltersRemoval` supposedly tests that the filter got removed
but, with `anyTree` here, we can't really know, can we?

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java [hunk 2]
```diff
@@ -254,8 +261,8 @@ public void testNestedDynamicFiltersRemoval()
                                 INNER,
                                 ImmutableList.of(equiJoinClause("ORDERS_CK", "ORDERS_CK6")),
                                 ImmutableMap.of("ORDERS_CK", "ORDERS_CK6"),
-                                Optional.empty(),
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:257]: I think this test requires some comment. As I read it now, I'm not sure what nested DF this test is about.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java [hunk 3]
```diff
@@ -101,8 +103,8 @@ public void testJoin()
                                 INNER,
                                 ImmutableList.of(equiJoinClause("ORDERS_OK", "LINEITEM_OK")),
                                 ImmutableMap.of("ORDERS_OK", "LINEITEM_OK"),
-                                Optional.empty(),
-                                tableScan("orders", ImmutableMap.of("ORDERS_OK", "orderkey")),
+                                anyTree(
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:106]: nit: we could use `DynamicFilters#createDynamicFilterExpression` as in  `TestRemoveUnsupportedDynamicFilters#testDynamicFilterConsumedOnBuildSide` to be little more explicit

At least please validate that there is `filter` node above table scan

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java [hunk 4]
```diff
@@ -72,6 +72,7 @@ public void testNonInnerJoin()
                         join(
```
Comment (sopel39) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:72]: could we enhance `io.prestosql.sql.planner.optimizations.TestRemoveUnsupportedDynamicFilters` tests as part of this PR. Currently, these tests don't validate that joins no longer have DF.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java (hunk 1 above)
Comment (raunaqmorarka) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:264]: In `JoinMatcher#matchDynamicFilters` we perform a traversal on left side of join looking for dynamic filters in FilterNode. So in all the tests in this file even though `anyTree` is given on probe side, `matchDynamicFilters` should fail when expectations around presence of dynamic filter are violated.
But if it's clearer to explicitly check for FilterNode, I'll make the change to add that.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java (hunk 3 above)
Comment (raunaqmorarka) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:106]: If the filter is not there then `JoinMatcher#matchDynamicFilters` should fail. The semi-join DF tests are also on the same lines. I can add the FilterNode explicitly here if that's better.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java (hunk 4 above)
Comment (raunaqmorarka) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:72]: done

File: presto-main/src/test/java/io/prestosql/sql/planner/assertions/JoinMatcher.java [hunk 5]
```diff
@@ -116,11 +131,48 @@ public MatchResult detailMatches(PlanNode node, StatsProvider stats, Session ses
...
+        return new MatchResult(matchDynamicFilters(joinNode, symbolAliases));
+    }
+
+    private boolean matchDynamicFilters(JoinNode joinNode, SymbolAliases symbolAliases)
+    {
+        if (expectedDynamicFilterAliases.isEmpty()) {
+            return true;
+        }
+        Set<DynamicFilterId> dynamicFilterIds = joinNode.getDynamicFilters().keySet();
+        Multimap<DynamicFilterId, Symbol> idToProbeSymbolMultiMap = searchFrom(joinNode.getLeft())
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply presto-main/src/test/java/io/prestosql/sql/planner/assertions/JoinMatcher.java]: Yes, that's cleaner approach, I've made this change now

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java (hunk 3 above)
Comment (raunaqmorarka) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:106]: Also, using `createDynamicFilterExpression` would require knowing the DynamicFilterId. We are able to do that there as the PlanNode is manually created in those tests.

File: presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java (hunk 2 above)
Comment (raunaqmorarka) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:257]: This was done to fix https://github.com/prestosql/presto/issues/2355
We made changes to remove DFs which became part of nested expressions due to predicate pushdown.

Comment (sopel39) [reply presto-main/src/test/java/io/prestosql/sql/planner/TestDynamicFilter.java:257]: How does it become part of nested expression? I don't see this.

**Review by sopel39 — APPROVED** [review]
lgtm @ small comment

**Comment by sopel39** [comment]
merged, thanks!

**Pull request merged**